### ISSUE TITLE
StringMatcher

### ISSUE THREAD

**Issue by orionll** [google/guava#2692]
We have a `CharMatcher` but its capabilities are limited. I would like to have the ability to match more complex strings. Examples:

- Strings starting with three letters and ending with a digit (e.g. `"var3"`, `"abc8"`):
`StringMatcher.inRange('a', 'z').times(3).then(StringMatcher.digit())`
- Strings containing words with semicolons (e.g. `"cat;dog;bird;"`):
`StringMatcher.inRange('a', 'z').oneOrMore().then(StringMatcher.char(';')).oneOrMore()`
- Strings representing integer numbers (e.g. `"17"`, `"-1"`):
`StringMatcher.anyOf("-+").optional().then(StringMatcher.digit().oneOrMore())`
- Strings containing either words or `123` (e.g. "`bird`", `"dDd"`, `"123"`):
`StringMatcher.inRange('a', 'z').or(StringMatcher.inRange('A','Z')).oneOrMore()
   .or(StringMatcher.str("123"))`

Yes, these all can be done via regular expressions. But I hate regular expressions since they are not readable and not checked at compile time.

Do we need such feature in Guava?

**Comment by orionll** [comment]
Of course, we don't need not turn Guava into a powerful parser combinator library. But having some basic capabilities of matching strings would be nice. I think, matching simple regular grammars will be enough.

**Comment by ineuwirth** [comment]
Hi @orionll 
What about https://github.com/VerbalExpressions/JavaVerbalExpressions?
Some more complex examples: https://github.com/VerbalExpressions/JavaVerbalExpressions/wiki/Parse-long-strings-example

**Comment by orionll** [comment]
> What about https://github.com/VerbalExpressions/JavaVerbalExpressions?
The last commit was 6 Apr. Looks dead.

**Comment by Zefick** [comment]
Regular expressions is the foundation while any third-party solution becomes "dead" with no commits in half of a year. There are a lot of online-testers for regexes which helps to debug an expressions. IDEs have either an integrated checkers (IDEA) or allow to install a plugins. Any way you should test your code despite its readability.

**Comment by ogregoire** [comment]
I don't know if it's because I've played enough with regexes to know them, but I find `[a-z]{3}\d` (or `[a-z]{3}[0-9]`) much more readable than `StringMatcher.inRange('a', 'z').times(3).then(StringMatcher.digit())`.

**Comment by orionll** [comment]
The problem with regexes is that when you need something more complex than `[a-z]{3}\d`, you get a long unreadable expression with a lot of escaped characters and weird control symbols which I have to google every time to know what they mean.

Another major problem with regexes is that you cannot give names to patterns inside your expression in order to reuse that pattern by name. For example, if you need to match two IP-addresses separated by a comma, you have to construct an expression "<long IP-address regex>,<long IP-address regex>" whereas in Java you can just write:
```
StringMatcher ip = <long IP-address matcher>;
ip.then(character(',')).then(ip);
```

**Comment by ogregoire** [comment]
String ipRegex =<long IP-address regex>
    String regex = ipRegex + "," + ipRegex;

I don't see the problem here.

**Comment by orionll** [comment]
Yes, my colleague wrote such kind of code (he was fired).

**Comment by ogregoire** [comment]
> Yes, my colleague wrote such kind of code (he was fired).

So? What's your point? I don't see any link between the two statements nor do I see how relevant they are to the discussion here.

**Comment by orionll** [comment]
The problem is clear. If you need to match `"["` instead of `","`, you will get a runtime error:
```
String ipRegex = <long IP-address regex>
String regex = ipRegex + "[" + ipRegex; // IDEA doesn't highlight this error
```

**Comment by ogregoire** [comment]
Then you come back to @Zefick's answer: use regex debuggers (which IDEA includes, according to Zefick, I don't use IDEA), unit-testing, etc.

**Comment by orionll** [comment]
Please, don't tell me what to use, OK? I will use what I prefer.

**Comment by jbduncan** [comment]
I've posted my own opinion about having a class like this in Guava at https://github.com/google/guava/issues/1075#issuecomment-269704845. **tl;dr:** I think this should seriously be considered for inclusion (or some variant of it at least) when Guava 21.0 (and 22.0?) are out of the door.

**Comment by orionll** [comment]
Thanks @jbduncan. I'm happy to know that there people who care about the type safety and the quality of their software.

**Comment by Stephan202** [comment]
@orionll, you seem to imply that people advocating for "just" using regular expressions don't "care about the type safety and the quality of their software". That's kind of a blanket statement (or even a covert insult).

Note that the type safety provided by the proposed alternative would guard against _syntax_ errors, but not necessarily semantic errors. The question is then: which category of error is the user of regular expressions more likely to make. The latter, I claim, because a dev will run the code they write at least once, thus uncovering syntax errors before they're ever committed.

As for software quality: regular expressions are well understood and should be part of every developer's toolkit. To invent an alternative could thus also be considered a case of [NIH](https://en.wikipedia.org/wiki/Not_invented_here).

**Comment by cgdecker** [comment]
I'm pretty much in agreement with people saying that you should use regex for this, but I'll leave it open because it's not impossible the idea could go somewhere at some point. I'd consider it very unlikely though.
